File: src-tauri/src/commands/ai.rs
//! Tauri commands for AI operations

use crate::ai::{
    detect_lmstudio, detect_ollama_models, load_ai_config, save_ai_config, test_provider_connection,
    AIConfig, AIProvider, ProviderTestResult, ProviderType,
};
use crate::AppPool;
use tauri::State;

// ============================================================================
// AI Configuration
// ============================================================================

/// Get the current AI configuration
#[tauri::command]
pub async fn get_ai_config(pool: State<'_, AppPool>) -> Result<AIConfig, String> {
    let pool_guard = pool.0.read().unwrap();
    let pool = pool_guard.as_ref().ok_or("Database not initialized")?;
    let conn = pool.get().map_err(|e| format!("Database error: {}", e))?;
    load_ai_config(&conn)
}

/// Save AI configuration
#[tauri::command]
pub async fn save_ai_config_cmd(
    pool: State<'_, AppPool>,
    config: AIConfig,
) -> Result<(), String> {
    let pool_guard = pool.0.read().unwrap();
    let pool = pool_guard.as_ref().ok_or("Database not initialized")?;
    let conn = pool.get().map_err(|e| format!("Database error: {}", e))?;
    save_ai_config(&conn, &config)?;
    Ok(())
}

/// Update a single provider's configuration
#[tauri::command]
pub async fn update_provider(
    pool: State<'_, AppPool>,
    provider: AIProvider,
) -> Result<AIConfig, String> {
    let pool_guard = pool.0.read().unwrap();
    let pool = pool_guard.as_ref().ok_or("Database not initialized")?;
    let conn = pool.get().map_err(|e| format!("Database error: {}", e))?;
    let mut config = load_ai_config(&conn)?;

    // Find and update the provider
    if let Some(existing) = config.providers.iter_mut().find(|p| p.id == provider.id) {
        *existing = provider;
    } else {
        config.providers.push(provider);
    }

    save_ai_config(&conn, &config)?;
    Ok(config)
}

/// Set the default provider
#[tauri::command]
pub async fn set_default_provider(
    pool: State<'_, AppPool>,
    provider_id: String,
) -> Result<AIConfig, String> {
    let pool_guard = pool.0.read().unwrap();
    let pool = pool_guard.as_ref().ok_or("Database not initialized")?;
    let conn = pool.get().map_err(|e| format!("Database error: {}", e))?;
    let mut config = load_ai_config(&conn)?;

    // Verify the provider exists
    if !config.providers.iter().any(|p| p.id == provider_id) {
        return Err(format!("Provider '{}' not found", provider_id));
    }

    config.default_provider = Some(provider_id);
    save_ai_config(&conn, &config)?;
    Ok(config)
}

/// Apply AI configuration
#[tauri::command]
pub async fn apply_ai_config(
    pool: State<'_, AppPool>,
) -> Result<(), String> {
    // Configuration is applied directly when making LLM calls
    let pool_guard = pool.0.read().unwrap();
    let _pool = pool_guard.as_ref().ok_or("Database not initialized")?;
    Ok(())
}

/// Initialize AI configuration
#[tauri::command]
pub async fn init_ai_config_cmd(pool: State<'_, AppPool>) -> Result<AIConfig, String> {
    let pool_guard = pool.0.read().unwrap();
    let pool = pool_guard.as_ref().ok_or("Database not initialized")?;
    let conn = pool.get().map_err(|e| format!("Database error: {}", e))?;
    crate::ai::init_ai_config(&conn)
}

// ============================================================================
// Provider Testing
// ============================================================================

/// Test a provider's API connection
#[tauri::command]
pub async fn test_provider(provider: AIProvider) -> ProviderTestResult {
    test_provider_connection(&provider, "").await
}

// ============================================================================
// Local Model Detection
// ============================================================================

/// Detect available local models from Ollama and LM Studio
#[tauri::command]
pub async fn detect_local_models(
    pool: State<'_, AppPool>,
) -> Result<AIConfig, String> {
    // Get pool and load config (sync work before async)
    let (db_pool, mut config, ollama_url, lmstudio_url) = {
        let pool_guard = pool.0.read().unwrap();
        let pool = pool_guard.as_ref().ok_or("Database not initialized")?.clone();
        let conn = pool.get().map_err(|e| format!("Database error: {}", e))?;
        let config = load_ai_config(&conn)?;
        
        let ollama_url = config.providers.iter()
            .find(|p| p.id == "ollama")
            .and_then(|p| p.base_url.clone())
            .unwrap_or_else(|| "http://localhost:11434".to_string());
        
        let lmstudio_url = config.providers.iter()
            .find(|p| p.id == "lmstudio")
            .and_then(|p| p.base_url.clone())
            .unwrap_or_else(|| "http://localhost:1234/v1".to_string());
        
        (pool, config, ollama_url, lmstudio_url)
    };

    // Detect Ollama models (async)
    let ollama_result = detect_ollama_models(&ollama_url).await;
    if ollama_result.success {
        if let Some(models) = ollama_result.models {
            log::info!("Detected {} Ollama models", models.len());
            if let Some(ollama) = config.providers.iter_mut().find(|p| p.id == "ollama") {
                ollama.models = models;
            }
        }
    } else {
        log::warn!("Failed to detect Ollama models: {}", ollama_result.message);
    }

    // Detect LM Studio models (async)
    let lmstudio_result = detect_lmstudio(&lmstudio_url).await;
    if lmstudio_result.success {
        if let Some(models) = lmstudio_result.models {
            log::info!("Detected {} LM Studio models", models.len());
            if let Some(lmstudio) = config.providers.iter_mut().find(|p| p.id == "lmstudio") {
                lmstudio.models = models;
            }
        }
    } else {
        log::warn!("Failed to detect LM Studio models: {}", lmstudio_result.message);
    }

    // Save updated config (sync work after async)
    {
        let conn = db_pool.get().map_err(|e| format!("Database error: {}", e))?;
        save_ai_config(&conn, &config)?;
    }
    
    Ok(config)
}

/// Detect Ollama installation and models
#[tauri::command]
pub async fn detect_ollama(base_url: Option<String>) -> ProviderTestResult {
    let url = base_url.as_deref().unwrap_or("http://localhost:11434");
    detect_ollama_models(url).await
}

/// Detect LM Studio models
#[tauri::command]
pub async fn detect_lmstudio_cmd(base_url: Option<String>) -> ProviderTestResult {
    let url = base_url.as_deref().unwrap_or("http://localhost:1234/v1");
    detect_lmstudio(url).await
}

// ============================================================================
// Provider Info
// ============================================================================

/// Get default provider configurations for all supported providers
#[tauri::command]
pub fn get_default_providers() -> Vec<AIProvider> {
    AIConfig::new().providers
}

/// Get provider type display info
#[tauri::command]
pub fn get_provider_info(provider_type: ProviderType) -> ProviderInfo {
    match provider_type {
        ProviderType::OpenAI => ProviderInfo {
            name: "OpenAI".to_string(),
            requires_api_key: true,
            default_base_url: None,
            description: "GPT-4, GPT-4o, and other OpenAI models".to_string(),
        },
        ProviderType::Anthropic => ProviderInfo {
            name: "Anthropic".to_string(),
            requires_api_key: true,
            default_base_url: None,
            description: "Claude 3.5 Sonnet, Haiku, and Opus models".to_string(),
        },
        ProviderType::Google => ProviderInfo {
            name: "Google".to_string(),
            requires_api_key: true,
            default_base_url: None,
            description: "Gemini Pro and Gemini Flash models".to_string(),
        },
        ProviderType::Ollama => ProviderInfo {
            name: "Ollama".to_string(),
            requires_api_key: false,
            default_base_url: Some("http://localhost:11434".to_string()),
            description: "Run open-source models locally with Ollama".to_string(),
        },
        ProviderType::LMStudio => ProviderInfo {
            name: "LM Studio".to_string(),
            requires_api_key: false,
            default_base_url: Some("http://localhost:1234/v1".to_string()),
            description: "Run models locally with LM Studio".to_string(),
        },
        ProviderType::VLLM => ProviderInfo {
            name: "VLLM".to_string(),
            requires_api_key: false,
            default_base_url: Some("http://localhost:8000/v1".to_string()),
            description: "Run models locally with VLLM server".to_string(),
        },
        ProviderType::Custom => ProviderInfo {
            name: "Custom".to_string(),
            requires_api_key: true,
            default_base_url: None,
            description: "Custom OpenAI-compatible endpoint".to_string(),
        },
    }
}

/// Information about a provider type
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderInfo {
    pub name: String,
    pub requires_api_key: bool,
    pub default_base_url: Option<String>,
    pub description: String,
}

File: src/types/ai.ts
/**
 * AI provider types
 */

/** Type of AI provider */
export type ProviderType =
  | "openai"
  | "anthropic"
  | "google"
  | "ollama"
  | "lmstudio"
  | "vllm"
  | "custom";

/** Configuration for a single AI provider */
export interface AIProvider {
  /** Unique identifier for this provider config */
  id: string;
  /** Display name */
  name: string;
  /** Provider type */
  type: ProviderType;
  /** API key (for cloud providers) */
  apiKey?: string;
  /** Base URL for the provider API */
  baseUrl?: string;
  /** Whether this provider is enabled */
  isEnabled: boolean;
  /** List of available models for this provider */
  models: string[];
  /** Currently selected model for this provider */
  selectedModel?: string;
  /** Context window size in tokens (optional, uses provider defaults if not set) */
  contextLength?: number;
}

/** Default context lengths by provider type (in tokens) */
export const DEFAULT_CONTEXT_LENGTHS: Record<ProviderType, number> = {
  openai: 200_000,
  anthropic: 200_000,
  google: 1_000_000,
  ollama: 32_000,
  lmstudio: 32_000,
  vllm: 32_000,
  custom: 32_000,
};

/** Get the effective context length for a provider */
export function getEffectiveContextLength(provider: AIProvider): number {
  return provider.contextLength ?? DEFAULT_CONTEXT_LENGTHS[provider.type as ProviderType] ?? 32_000;
}

/** Configuration for embedding generation */
export interface EmbeddingConfig {
  /** The provider ID to use for embeddings (e.g., "ollama", "openai", "lmstudio") */
  provider: string;
  /** The embedding model to use (e.g., "nomic-embed-text" for ollama) */
  model: string;
  /** Dimension of the embedding vector */
  dimension: number;
  /** Whether to automatically embed notes on create/update */
  autoEmbed: boolean;
}

/** Helper to get the full model ID for API calls (provider/model format) */
export function getFullEmbeddingModelId(config: EmbeddingConfig): string {
  return `${config.provider}/${config.model}`;
}

/** Complete AI configuration */
export interface AIConfig {
  /** List of configured providers */
  providers: AIProvider[];
  /** ID of the default provider to use */
  defaultProvider: string | null;
  /** Embedding configuration */
  embedding: EmbeddingConfig;
}

/** Result of testing a provider connection */
export interface ProviderTestResult {
  success: boolean;
  message: string;
  models?: string[];
}

/** Result of local model detection */
export interface LocalModelsResult {
  ollama: ProviderTestResult;
  lmstudio: ProviderTestResult;
}

/** Information about a provider type */
export interface ProviderInfo {
  name: string;
  requiresApiKey: boolean;
  defaultBaseUrl: string | null;
  description: string;
}

/** Input for updating a provider */
export type UpdateProviderInput = Partial<Omit<AIProvider, "id">> & {
  id: string;
};

/** Default configuration for each provider type */
export const PROVIDER_DEFAULTS: Record<
  ProviderType,
  { name: string; requiresApiKey: boolean; defaultBaseUrl?: string }
> = {
  openai: { name: "OpenAI", requiresApiKey: true },
  anthropic: { name: "Anthropic", requiresApiKey: true },
  google: { name: "Google", requiresApiKey: true },
  ollama: {
    name: "Ollama",
    requiresApiKey: false,
    defaultBaseUrl: "http://localhost:11434",
  },
  lmstudio: {
    name: "LM Studio",
    requiresApiKey: false,
    defaultBaseUrl: "http://localhost:1234/v1",
  },
  vllm: {
    name: "VLLM",
    requiresApiKey: false,
    defaultBaseUrl: "http://localhost:8000",
  },
  custom: { name: "Custom", requiresApiKey: true },
};

/** Get display name for provider type */
export function getProviderDisplayName(type: ProviderType): string {
  return PROVIDER_DEFAULTS[type]?.name ?? type;
}

/** Check if provider type requires API key */
export function providerRequiresApiKey(type: ProviderType): boolean {
  return PROVIDER_DEFAULTS[type]?.requiresApiKey ?? true;
}

/** Check if provider type is a local provider */
export function isLocalProvider(type: ProviderType): boolean {
  return type === "ollama" || type === "lmstudio" || type === "vllm";
}
